### PR TITLE
Add support for array of types in schema definition

### LINKED ISSUE #77: Support for array of types in JSON schema
The following is a valid schema, but currently unsupported by fast-json-stringify:
```json
{
    "type": "object",
    "properties": {
      "stringOrNumber": {"type": ["string", "number"], "maxLength": 10}
    }
}
```

### PR DESCRIPTION
**Pull request by florianreinhart**
Fixes #77.

This adds support for `type` to be an array of types, e.g.

```json
{
    "type": "object",
    "properties": {
      "stringOrNumber": {"type": ["string", "number"]}
    }
}
```

Support of multiple types in tuple definitions is still missing. This probably requires some refactoring of the `buildArray()` function. Nevertheless, I have prepared already a test case for tuples (currently commented out).

I don't particularly like the implementation of `hasArrayOfTypes()`, so please let me know if you have any better ideas.

### CODE REVIEW

**Comment by allevo** [comment]
Many thanks for this PR. Can you run some benchmarks ? I like to see if there're some performance penalties between your branch and master.

**Comment by florianreinhart** [comment]
Sure! I ran `bench.js`. Here are the results:

**Before**
```
JSON.stringify array x 4,509 ops/sec ±0.70% (92 runs sampled)
fast-json-stringify array x 8,331 ops/sec ±0.68% (92 runs sampled)
fast-json-stringify-uglified array x 8,463 ops/sec ±0.75% (90 runs sampled)
JSON.stringify long string x 17,575 ops/sec ±0.46% (97 runs sampled)
fast-json-stringify long string x 17,035 ops/sec ±1.56% (92 runs sampled)
fast-json-stringify-uglified long string x 17,522 ops/sec ±0.50% (93 runs sampled)
JSON.stringify short string x 5,653,901 ops/sec ±0.46% (94 runs sampled)
fast-json-stringify short string x 32,673,355 ops/sec ±0.38% (94 runs sampled)
fast-json-stringify-uglified short string x 33,049,197 ops/sec ±0.30% (94 runs sampled)
JSON.stringify obj x 2,126,520 ops/sec ±0.62% (96 runs sampled)
fast-json-stringify obj x 7,336,898 ops/sec ±0.95% (95 runs sampled)
fast-json-stringify-uglified obj x 7,225,950 ops/sec ±0.95% (92 runs sampled)
```

**After**
```
JSON.stringify array x 4,567 ops/sec ±0.47% (92 runs sampled)
fast-json-stringify array x 8,327 ops/sec ±0.77% (93 runs sampled)
fast-json-stringify-uglified array x 8,660 ops/sec ±0.64% (94 runs sampled)
JSON.stringify long string x 17,648 ops/sec ±0.43% (97 runs sampled)
fast-json-stringify long string x 17,591 ops/sec ±0.20% (97 runs sampled)
fast-json-stringify-uglified long string x 17,170 ops/sec ±0.99% (93 runs sampled)
JSON.stringify short string x 5,741,197 ops/sec ±0.21% (96 runs sampled)
fast-json-stringify short string x 33,161,864 ops/sec ±0.53% (93 runs sampled)
fast-json-stringify-uglified short string x 32,049,135 ops/sec ±0.48% (94 runs sampled)
JSON.stringify obj x 2,145,600 ops/sec ±0.58% (93 runs sampled)
fast-json-stringify obj x 7,341,208 ops/sec ±0.61% (93 runs sampled)
fast-json-stringify-uglified obj x 7,281,431 ops/sec ±0.83% (91 runs sampled)
```

Looks OK to me.

If your schema doesn't contain any type arrays, the generated code is the same as before. However, the code generation itself could be a little slower.

File: test/typesArray.test.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+//   try {
+//     const value = stringify({
+//       fixedTupleOfStringsAndNumbers: ['string1', 42, 7]
+//     })
+//     t.is(value, '{"fixedTupleOfStringsAndNumbers":["string1",42,"string2"]}')
+//   } catch (e) {
+//     console.log(e)
+//     t.fail()
+//   }
+// })
```
Comment (mcollina) [inline test/typesArray.test.js]: Do you think it would be feasible to also implement this feature? If not, we should at least throw a nice error that this is not supported and test that.

File: index.js [hunk 2]
```diff
@@ -517,6 +551,7 @@ function buildArray (schema, code, name, externalSchema, fullSchema) {
           condition += `Array.isArray(obj${accessor})`
           break
         default:
+          // TODO: item.type can be an array of types.
```
Comment (mcollina) [inline index.js]: is this todo related to the commented text?

**Comment by mcollina** [comment]
Good work!

File: index.js (hunk 2 above)
Comment (florianreinhart) [reply index.js]: Yes. At this point we need to check if `item.type` is an array and then call into the switch statement from line 531 again for each element in the array. The switch statement should be factored out into a separate function.

**Comment by florianreinhart** [comment]
@mcollina I've now also added support for multiple types in tuples.

File: index.js [hunk 3]
```diff
@@ -561,6 +570,43 @@ function buildArray (schema, code, name, externalSchema, fullSchema) {
...
+      break
+    case 'array':
+      condition = `Array.isArray(obj${accessor})`
+      break
+    default:
+      if (Array.isArray(type)) {
+        var conditions = type.map((subType) => {
+          return buildArrayTypeCondition(subType, accessor)
+        })
+        condition = `(${conditions.join(' || ')})`
```
Comment (florianreinhart) [inline index.js]: There is potential for an optimization here. In the current implementation the generated code checks if the item matches any of the types. The `nested()` basically performs the same check again using ajv. We could split this condition here into separate ones and use the knowledge of the item's type for the `nested()` function. However, this requires some more refactoring.

Comment (delvedor) [reply index.js]: Feel free to experiment! We do like optimizations! :D

**Comment by florianreinhart** [comment]
Same machine and node v8.9.4 as yesterday. Looks OK.
```
JSON.stringify array x 4,568 ops/sec ±0.72% (96 runs sampled)
fast-json-stringify array x 8,553 ops/sec ±1.13% (91 runs sampled)
fast-json-stringify-uglified array x 8,594 ops/sec ±0.62% (92 runs sampled)
JSON.stringify long string x 17,850 ops/sec ±0.50% (94 runs sampled)
fast-json-stringify long string x 17,881 ops/sec ±0.17% (97 runs sampled)
fast-json-stringify-uglified long string x 17,849 ops/sec ±0.56% (97 runs sampled)
JSON.stringify short string x 5,813,338 ops/sec ±0.23% (94 runs sampled)
fast-json-stringify short string x 33,805,913 ops/sec ±0.45% (95 runs sampled)
fast-json-stringify-uglified short string x 33,736,264 ops/sec ±0.31% (96 runs sampled)
JSON.stringify obj x 2,180,008 ops/sec ±0.19% (97 runs sampled)
fast-json-stringify obj x 7,575,110 ops/sec ±0.79% (94 runs sampled)
fast-json-stringify-uglified obj x 6,840,425 ops/sec ±1.36% (89 runs sampled)
```

File: index.js (hunk 3 above)
Comment (mcollina) [reply index.js]: I'm landing this now, feel free to send a followup PR with the optimizations!
